### ISSUE TITLE
Semantically kill/yank citations between formats

### ISSUE THREAD

**Issue by Hugo-Heagren** [emacs-citar/citar#635]
**Is your feature request related to a problem? Please describe.**

I plan essays and write notes in org mode, and use org mode's citation syntax to plan my references.

But then I *write* in LaTeX. I have an org plan buffer open on the left, and my prose on the right. At the moment I have to manually look at the left-hand citation, and write out (or use citar to insert) the same citation again, in a different format.

**Describe the solution you'd like**

Tools like citar help a lot, but it would be great if I could just copy my citation from org mode (maybe with a citar command) and paste it into the LaTeX buffer *as* LaTeX code. A sort of semantic copy/paste. 

Of course, citar handles many different formats, so ideally this should just happen automagically (just like insertion does) -- you yank and the yanked text just looks like it should, in any buffer (including those not supported, where a simple text string is inserted).

**Describe alternatives you've considered**

Implementing a janky solution in my init file; processing manually as above.

**Additional context**

This could be implemented as part of `citar-copy-reference`, by attaching `yank-handler` and `citar-keys-entries` properties to the text (see `yank`'s docstring). The yank-handler function would check if we're in a buffer which citar can handle (tex, org mode, markdown etc.). If not, insert the text as-is. If so, call `citar-major-mode-function` with the value of the `keys-entries` property. (there are other details to work out, like how to undo this, but it's a start).

I would happy to try implementing this.

**Comment by bdarcus** [comment]
Interesting idea.

> ... it would be great if I could just copy my citation from org mode (maybe with a citar command) and paste it into the LaTeX buffer as LaTeX code. A sort of semantic copy/paste.

I wonder how common this use case is?

Don't most users tend to use one major for authoring documents with citations?

That's the only real concern I have; not really an objection.

So it probably depends on what the implementation looks like; whether there's any downside to it.

> This could be implemented as part of citar-copy-reference, by attaching yank-handler and citar-keys-entries properties to the text (see yank's docstring). 

Two things here:

1. wouldn't it be a `citar-copy-citation`? `citar-org` actually has such a function, and some related ones for modifying citations in the buffer (see #385).
2. note the breaking changes we're working on in #634. That will change the API to use citekeys.

**Comment by Hugo-Heagren** [comment]
> That's the only real concern I have; not really an objection.
> 
> So it probably depends on what the implementation looks like; whether there's any downside to it.

Fair point. I can imagine that some users might also prefer the simple text copy/yank. Perhaps the 'semantic' behaviour should be toggle-able with custom option? Something like `citar-use-mode-aware-yank`. The yank handler could just behave like `insert` if this option was `nil`.

> `citar-copy-citation`

This sounds like it makes sense, but I can't find any references to this function anywhere. 

> note the breaking changes we're working on

Ahh. Well, I was only raising an idea, and it would be quite a while until I had time to look into anyway, so I might wait for that PR to stabilise.

**Comment by bdarcus** [comment]
> This sounds like it makes sense, but I can't find any references to this function anywhere? Have I misunderstood and you're suggesting this as a name for the function?

Yeah, I was suggesting the possibility of one or more new major-mode functions.

**Comment by aikrahguzar** [comment]
> Yeah, I was suggesting the possibility of one or more new major-mode functions.

Why does it need a new major mode function? To me it seems like on one side we need `citation-at-point` to extract the keys and on the other side `insert-citation`. Copying part of the operation will likely need a new command but pasting should be doable as normal using a `yank-handler` property on the keys.

The only thing missing is citation command but doesn't seem like it is doable using an addition to major-mode-functions since it needs information from two major-modes and seems something that is hopeless to do in general. I think the best thing to do is to as for a normal citation and either a default cite command or prompt for one.

**Comment by bdarcus** [comment]
> I think the best thing to do is to as for a normal citation and either a default cite command or prompt for one.

Can you imagine an elegant UX for that?

If most users only copy and paste within the same document, or major mode, we wouldn't want to both them to confirm that.

Is it possible to prompt if the major mode changes?

**Comment by aikrahguzar** [comment]
> Can you imagine an elegant UX for that?
> 
> If most users only copy and paste within the same document, or major mode, we wouldn't want to both them to confirm that.

I think copy and pasting within the same  mode is an editing operation and I would prefer to use something that is part of normal editing experience of a mode. In latex for example I would use `evil-tex-a-command` text operator to do this.

But ignoring that, this doesn't seem too difficult. From the documentation of yank
```
The value of a yank-handler property must be a list of one to four
elements, of the form (FUNCTION PARAM NOEXCLUDE UNDO).
FUNCTION, if non-nil, should be a function of one argument (the
 object to insert); FUNCTION is called instead of insert.
PARAM, if present and non-nil, is passed to FUNCTION (to be handled
 in whatever way is appropriate; e.g. if FUNCTION is yank-rectangle,
 PARAM may be a list of strings to insert as a rectangle)
```
So the major mode from which the citation was copied can be used as the `PARAM` and then compared against the `major-mode` of the target buffer.

If someone on a recent enough emacs has a creative idea of how to detect when a string is a citation, they can even use `kill-transform-function` to do this propertization automatically but it definitely feels something more suitable in a config than a package.

**Comment by Hugo-Heagren** [comment]
> So the major mode from which the citation was copied can be used as the `PARAM` and then compared against the `major-mode` of the target buffer.

Yes, this would work. I was thinking of a slightly different way: that on copying, the citation string could be parsed, and an internal mode-agnostic representation stored in a text property. Then the yank handler just calls `citar--major-mode-function` with that value (`citar--major-mode-function` being a function already built to do detection and handling of major modes). Same result, different mechanics :shrug:
